init: add sky42 enhancements/bug fixes

MilhouseVH opened the pull request:
As discussed on the forum:

1) Upgrade from img.gz fails when using a custom kernel filename ([thread](https://forum.libreelec.tv/thread/12400-update-with-new-le-img-gz-failes-when-boot-image-kern-82))
2) /flash/EFI/BOOT/syslinux.cfg is redundant ([thread](https://forum.libreelec.tv/thread/12411-just-one-syslinux-cfg-to-edit-for-bios-uefi))
3) Boot hooks (for encryption, other things) and configurable system to ram limit ([thread](https://forum.libreelec.tv/thread/12413-2-init-boot-hooks-and-kernel-argument-to-configure-system-toram-limit-patch-incl))

Thanks sky42.

vpeter4 commented on packages/sysutils/busybox/scripts/init:
```diff
@@ -90,7 +93,10 @@
     case $arg in
       BOOT_IMAGE=*)
-        IMAGE_KERNEL="${arg#*=}"
+        if [ "${BOOT_IMAGE_DONE}" = "no" ]; then
```
Why do you need another variable to check? Just use

```
        if [ "${IMAGE_KERNEL}" = "KERNEL" ]; then
          IMAGE_KERNEL="${arg#*=}"
        fi
```

vpeter4 replied:
Ah, then it is no go :)
Seems variable will suffice.

MilhouseVH commented:
Just to spell out the BOOT_IMAGE issue, because it's not easy/obvious:

The key issue raised by sky42 on the forum is that we haven't correctly handled the mapping when dealing with an upgrade from an img.

Using the RPi as an example, these are the default filenames in a tar and img:
```
tar file: KERNEL       SYSTEM
img file  kernel.img   SYSTEM
```

When booting with a custom `BOOT_IMAGE=/kern82` we then try to find kern82 in the img file, which doesn't exist.

If using x86:
```
tar file: KERNEL       SYSTEM
img file  KERNEL       SYSTEM
```
and booting with custom `BOOT_IMAGE=/kern82` the upgrade also fails for the same reason as RPi.

During `init` we need to know two things that vary from device to device, and installation to installation:

1) The default kernel name in the img (varies from device-to-device, set in `init` by `sed "s/@KERNEL_NAME@/$KERNEL_NAME/"`)
2) The booted kernel name (varies from installation-to-installation, from `BOOT_IMAGE=`)

Fortunately the default kernel name in a tar is always KERNEL, which is one less issue to worry about.

When the kernel always appends the contents of `BOOT_IMAGE=@KERNEL_NAME@` on the command line this complicates this processing of a custom installation (#2, above), hence the hack.

To solve this, we can:

1) add a device specific file  that with `KERNEL_IMAGE=@KERNEL_NAME@`
2) continue using the `sed`, but now setting KERNEL_IMAGE="@KERNEL_NAME@" which should work (currently it is hard coded to use `"KERNEL"`)

I've added an extra commit which implements #2, as handling this with `sed` is a trivial change given existing code.

Now that we are reliably handling the different kernel names without a hard dependence on `BOOT_IMAGE=`, we can drop `BOOT_IMAGE=` from RPi/Rockchip kernels (it's not present in a Generic kernel, but syslinux will add it during boot and we can trust syslinux, so long as the user doesn't also add it - PEBKAC in that case).

I did consider dropping `BOOT_IMAGE=` from Amlogic kernels, but it's used elsewhere (in `platform_init`, `update.sh`) so I'm thinking we leave it in Amlogic kernels and deal with it later for mainline.

This all needs testing, of course... :)

mglae commented on packages/sysutils/busybox/scripts/init:
```diff
@@ -719,7 +719,6 @@
 
           mount -o remount,rw /flash
           sed -i "s/^DEFAULT .*/DEFAULT $SYSLINUX_DEFAULT/" /flash/syslinux.cfg
-          [ -f /flash/EFI/BOOT/syslinux.cfg ] && cp /flash/syslinux.cfg /flash/EFI/BOOT/syslinux.cfg
```
The line is still needed if a LE8 system is updated to LE9. 

Or remove any /flash/EFI/BOOT/syslinux.cfg here.

MilhouseVH replied:
Ah yes, you're right.

This code is only used by the `installer`/`live`/`run` modes (typically, when booting from a USB memory stick) and we won't have included `EFI/BOOT/syslinux.cfg` in such a build.

However, if someone is booting their PC via USB memory stick (or has dd'd the img directly to HDD disk) and is always using `run` mode to boot, then once they upgrade from LE8 the new LE9 `init` will stop overwriting their EFI config, and they would "lose" the ability to remember the last boot mode (assuming they ever booted into another mode).

I'll update the commit so that it removes `/flash/EFI/BOOT/syslinux.cfg` - thanks.

MilhouseVH commented:
Added a "MERGE BLOCKED" label while we wait for a reply from @maideii about grub configs - not essential but would be nice to nail that too if possible. Will remove in a couple of days.

maideii commented:
Hi, I am in field trip for a week if you guys can wait until I am back to my station then maybe I can start to check out the config. But in my understanding grub always using absolute path to find the grub.cfg where we install them. So using the existing installation config will not work if just moving the grub.cfg to /flash/syslinux.cfg.

MilhouseVH commented:
Hi @maideii - glad you're still around!

There's no immediate rush to merge this so next week or so should be fine.

> But in my understanding grub always using absolute path to find the grub.cfg where we install them. So using the existing installation config will not work if just moving the grub.cfg to /flash/syslinux.cfg.

I think you misunderstand - we're not moving `grub.cfg` to `/flash/syslinux.cfg`.

What I'm proposing is that we no longer install `/flash/EFI/BOOT/grub.cfg`, as it may be that `grub` will use `/flash/grub.cfg` whenever `/flash/EFI/BOOT/grub.cfg` does not exist, however I need to have this behaviour confirmed.

With `syslinux` we have been installing both `/flash/syslinux.cfg` and also `/flash/EFI/BOOT/syslinux.cfg`. It turns out that installing both `syslinux.cfg` files is unnecessary as `syslinux` will use `/flash/syslinux.cfg` whenever `/flash/EFI/BOOT/syslinux.cfg` doesn't exist. So with this PR we will no longer install `/flash/EFI/BOOT/syslinux.cfg`, and `syslinux` still boots with both UEFI and non-UEFI.

What I'm hoping you can confirm for me is whether `grub` behaves in a similar way to `syslinux`, and if so then hopefully we can stop installing `/flash/EFI/BOOT/grub.cfg`. If not, then we will leave the `grub` config files as they are.

Thanks!

MilhouseVH commented:
@maideii given the way we are building `grub` (with `/EFI/BOOT` as the prefix) it may be that we need `/flash/EFI/BOOT/grub.cfg` but do not need `/flash/grub.cfg`. If we need both files then no changes are necessary, but if we only need one `grub.cfg` file then we can stop installing the redundant file.

Are you able to test `grub` when booting with UEFI and without UEFI? Hopefully it's possible to test both. Thanks again.

maideii commented:
Here is the testing result:

In UEFI mode:
1) only grub.cfg in /flash/EFI/BOOT boot as expected (tested in all Install, Run and Live mode).
2) only grub.cfg in /flash will not boot and EFI will request the location of the configfile.

* It is safe to remove the grub.cfg at /flash for the current installation config but not in /flash/EFI/BOOT.

In legacy bios mode:
1) x64 pc will start syslinux properly.
2) i386/i686 won't start with syslinux.
3) no grub mode in i386 or x64.

MilhouseVH commented:
@maideii many thanks. Can you check https://github.com/LibreELEC/LibreELEC.tv/pull/2689/commits/8b774703ee7ce72fdc9dc2ad5bbd7ccf220b213f to make sure I've not misunderstood anything - this commit removes `/flash/grub.cfg` so that only `/flash/EFI/BOOT/grub.cfg` is installed.

@lrusak, @HiassofT, @vpeter4, @mglae I've dropped support for `extlinux` as this seems unnecessary - we can install `syslinux` regardless of UEFI boot mode as `syslinux` supports both UEFI and non-UEFI, whereas `extlinux` only  supports non-UEFI booting.

I've also updated the `installer` to use UUIDs for the installed system (`syslinux.cfg` and `grub.cfg`).

vpeter4 commented on packages/tools/installer/scripts/installer:
```diff
@@ -262,27 +262,26 @@ do_install_quick() {
-        echo "	search --set -f /KERNEL" >> $TMPDIR/part1/grub.cfg
-        echo "	linux /KERNEL boot=LABEL=$DISKLABEL_SYSTEM disk=LABEL=$DISKLABEL_STORAGE quiet" >> $TMPDIR/part1/grub.cfg
-        echo "}" >> $TMPDIR/part1/grub.cfg
+        echo "set timeout=\"0\"" > $TMPDIR/part1/EFI/BOOT/grub.cfg
```
You could use here multiline string using `EOF` block as you did on other places. The code is much more readable.

MilhouseVH replied:
I did - next commit. :)

vpeter4 replied:
Ah, there were to many changes to see that :)

MilhouseVH commented:
"Merge Blocked" label removed.
